### DISCUSSION TITLE
help - FormatConversion not working with stream

### DISCUSSION THREAD

**Discussion by MohammedSamy** [pschatzmann/arduino-audio-tools#785] [Q&A] [answered]
Hello,

I am doing a project where I have an ESP32 reciving I2S commuincation and sends that exact data to two I2S outputs but one of outputs would be delayed by a controllable amount (I want to sync my the audio between my TV and speakers).

So I am trying to takle the problems in sequence. While I was testing the delay effect, I think it can only work well on sample rate 44100 and 1 channel (mentioned in one of the disscusions). So I want to make sure I can get a working delay effect on only one of the outputs. But to test that I need a source I can test with that has 44100 sample rate. I am trying to use the stream mp3 example sketch, but I am not able to convert the output to sample rate 44100. The mp3 URL always forces the I2s to sample rate 44100.

- Would appricate some help how to get a streaming source with 44100 sample rate and 1 channel
- or help on how to convert to the desired sample rate. Here is my current sketch, I tried the format converter on every stage in the pipeline but always the copy function doesnt copy. Here is the error from the copy function:`
[D] AudioCopy.h : 90 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]`

**Sketch Code:**
````

I2SStream i2sStream1;

I2SStream i2s; // I2S output
MultiOutput out;
FormatConverterStream conv(out);
EncodedAudioStream out2dec(&out, new MP3DecoderHelix()); // Decoding stream
AudioEffectStream effects(i2sStream1); // apply effects to output: writing to effects



StreamCopy copier(out2dec, url); // copy url to decoder
Delay dly(200, 0.5, 1.0,44100, true);

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Debug);  
    Wire.begin();
    Wire1.begin(13,12);
    
  // Serial.println("I2C Scanner");
  while (0 != WM8960_Init())
    ;

   // WM8960_Volume(0.5,0.5);

        
  // Serial.println("I2C Scanner");
  while (0 != WM8960_Init_1())
    ;
     
     
      effects.addEffect(dly);

  // setup multi output  
  out.add(i2s);
  out.add(i2sStream1);

  // setup i2s
  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.i2s_format = I2S_STD_FORMAT; // if quality is bad change to I2S_LSB_FORMAT https://github.com/pschatzmann/arduino-audio-tools/issues/23
  cfg.sample_rate = 44100;
  cfg.channels = 1;
  cfg.bits_per_sample = 16;
    cfg.i2s_format = I2S_STD_FORMAT;
  cfg.pin_bck = PIN_I2S_BCK;
  cfg.pin_ws = PIN_I2S_WS;
 // cfg.pin_data_rx = PIN_I2S_DATA_IN;
  cfg.pin_data = 18;
  cfg.is_master = true;
  cfg.port_no = 0;
 // cfg.buffer_size = 1024;
  i2s.begin(cfg);



 Serial.println("starting I2S...");
  auto config1 = i2sStream1.defaultConfig(TX_MODE);
  config1.i2s_format = I2S_STD_FORMAT; // if quality is bad change to I2S_LSB_FORMAT https://github.com/pschatzmann/arduino-audio-tools/issues/23
  config1.sample_rate = 44100;
  config1.channels = 1;
 config1.bits_per_sample = 16;
    config1.i2s_format = I2S_STD_FORMAT;
  config1.pin_bck = 16;
  config1.pin_ws = 17;
  //config1.pin_data_rx = PIN_I2S_DATA_IN;
  config1.pin_data = 5;
  config1.port_no = 1;
    config1.is_master = true;
      //  config1.buffer_size = 1024;
//        config1.buffer_count=20;
  i2sStream1.begin(config1);
  Serial.println("I2S started");

AudioBaseInfo from;AudioBaseInfo to;
  from.channels=2;
  from.bits_per_sample=16;
  from.sample_rate=48000;

    to.channels=1;
  to.bits_per_sample=16;
  to.sample_rate=44100;

  conv.begin(from, to);

 auto config2= out2dec.defaultConfig();
  config2.channels=1;
  config2.bits_per_sample=16;
  config2.sample_rate=44100;
 
  // setup I2S based on sampling rate provided by decoder
//  out2dec.setNotifyAudioChange(i2s);
  out2dec.begin(config2);

  auto cffg = effects.defaultConfig();
  cffg.channels=1;
  cffg.bits_per_sample=16;
  cffg.sample_rate=44100;
effects.begin( cffg);
// mp3 radio

   url.begin("http://stream.srg-ssr.ch/m/rsj/mp3_128","audio/mp3");
 // sineWave.begin(1, 44100, N_B4);
}

void loop(){
  copier.copy();

}
```
`

**Comment by pschatzmann** [comment, marked as answer]
MP3 contains the sample rate, so the EncodedAudioStream automatically sets the audio information provided by the mp3.
I don't understand why you add this complexity when you plan to get the data from i2s in the end.
For testing just keep things simple and use a sine generator (or just use an mp3 file recorded in your desired sample rate!)

I also suggest to start simple and add complexity step, by step. That's how you will find where you made your error!

You can also try just to use the sample rate and i2s buffer to control the delay: the smaller the rate the bigger the delay and the bigger the buffer the bigger the delay.

**Comment by MohammedSamy** [comment]
`MP3 contains the sample rate, so the EncodedAudioStream automatically sets the audio information provided by the mp3. `
>>> Yes I figured that out, although took a few hours to do so. would you happen to know a way or a place where I can search for mp3 streaming websites with sample rate 44100?


```
I don't understand why you add this complexity when you plan to get the data from i2s in the end.
For testing just keep things simple and use a sine generator! 
```
>>>> I did try to use a sine generator, and it somewhat works, but I am not able to verify  that the delay is working, I do not have an oscoliscope unfortunatly, I am relying on an audible que.  


`I also suggest to start simple and add complexity step, by step. That's how you will find where you made your error!`
>>> yes agree, that is exactly what I am doing. I am trying to make sure I can do a reliable delay on an I2s output while the other does not have a delay. this is the point I am at right now.  I am trying to solve the delay problem before adding a 3rd I2S in the mix as an input. So I am trying to use the example sketches as much as possible.


`You can also try just to use the sample rate and i2s buffer to control the delay: the smaller the rate the bigger the delay and the bigger the buffer the bigger the delay.`
>>> Yes, I tried that based on your suggestion in the issue I mistakingly created. But everytime I set the I2S buffer more than 1024 I get a stack overflow error. I am guessing that too much ram is being used in this case, espically with the sample rate 44.1k. And unfortunalty I am locked to the sample rate because I am trying to create a sort of a homethreatre solution (from my limited reserch 44.1k is minimum to have a good audio solution)

**Reply by pschatzmann** [reply]
The delay is not working because you don't don't use it anywhere! 
You control the total i2s buffer size with the buffer_size and buffer_count
Also the conv is not used anywhere in your chain

**Reply by MohammedSamy** [reply]
yes, in this code I am not using the delay. I had it before and commented it out because I had artifacts in the sound when it was on.  So, I was trying to get a 44.1k sample rate source working to test teh delay effect on to see if the sound artifacts remain.

For the I2s buffer_size and buffer_count, yes I had tried them before (not in this sketch) so that they act as a delay. But every time I try to increase the I2S buffer_size to more than 1024 I get a stack overflow error.


I will try the option you suggested for recording an mp3, good Idea.  But is it possible, could you check the code what could be wrong with the formatconverter stream not working?

And super thanks by the way. Your work on this library and the moderation is absolutly amazing.

**Reply by pschatzmann** [reply]
The conv is not used anywhere in your chain

**Reply by MohammedSamy** [reply]
oh apologies, my mistake. I copied the wrong code. In the code I coped above I was making sure I can still output the stream to multioutput I2S ports.  Here is the code with the coonverter that when I use, the coper does not copy anything and produces the error mentioned above:

````

I2SStream i2sStream1;

I2SStream i2s; // I2S output
MultiOutput out;
FormatConverterStream conv(out);
EncodedAudioStream out2dec(&conv, new MP3DecoderHelix()); // Decoding stream
AudioEffectStream effects(i2sStream1); // apply effects to output: writing to effects



StreamCopy copier(out2dec, url); // copy url to decoder
Delay dly(200, 0.5, 1.0,44100, true);

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Debug);  
    Wire.begin();
    Wire1.begin(13,12);
    
  // Serial.println("I2C Scanner");
  while (0 != WM8960_Init())
    ;

   // WM8960_Volume(0.5,0.5);

        
  // Serial.println("I2C Scanner");
  while (0 != WM8960_Init_1())
    ;
     
     
      effects.addEffect(dly);

  // setup multi output  
  out.add(i2s);
  out.add(i2sStream1);

  // setup i2s
  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.i2s_format = I2S_STD_FORMAT; // if quality is bad change to I2S_LSB_FORMAT https://github.com/pschatzmann/arduino-audio-tools/issues/23
  cfg.sample_rate = 44100;
  cfg.channels = 1;
  cfg.bits_per_sample = 16;
    cfg.i2s_format = I2S_STD_FORMAT;
  cfg.pin_bck = PIN_I2S_BCK;
  cfg.pin_ws = PIN_I2S_WS;
 // cfg.pin_data_rx = PIN_I2S_DATA_IN;
  cfg.pin_data = 18;
  cfg.is_master = true;
  cfg.port_no = 0;
 // cfg.buffer_size = 1024;
  i2s.begin(cfg);



 Serial.println("starting I2S...");
  auto config1 = i2sStream1.defaultConfig(TX_MODE);
  config1.i2s_format = I2S_STD_FORMAT; // if quality is bad change to I2S_LSB_FORMAT https://github.com/pschatzmann/arduino-audio-tools/issues/23
  config1.sample_rate = 44100;
  config1.channels = 1;
 config1.bits_per_sample = 16;
    config1.i2s_format = I2S_STD_FORMAT;
  config1.pin_bck = 16;
  config1.pin_ws = 17;
  //config1.pin_data_rx = PIN_I2S_DATA_IN;
  config1.pin_data = 5;
  config1.port_no = 1;
    config1.is_master = true;
      //  config1.buffer_size = 1024;
//        config1.buffer_count=20;
  i2sStream1.begin(config1);
  Serial.println("I2S started");

AudioBaseInfo from;AudioBaseInfo to;
  from.channels=2;
  from.bits_per_sample=16;
  from.sample_rate=48000;

    to.channels=1;
  to.bits_per_sample=16;
  to.sample_rate=44100;

  conv.begin(from, to);

 auto config2= out2dec.defaultConfig();
  config2.channels=1;
  config2.bits_per_sample=16;
  config2.sample_rate=44100;
 
  // setup I2S based on sampling rate provided by decoder
//  out2dec.setNotifyAudioChange(i2s);
  out2dec.begin(config2);

  auto cffg = effects.defaultConfig();
  cffg.channels=1;
  cffg.bits_per_sample=16;
  cffg.sample_rate=44100;
effects.begin( cffg);
// mp3 radio

   url.begin("http://stream.srg-ssr.ch/m/rsj/mp3_128","audio/mp3");
 // sineWave.begin(1, 44100, N_B4);
}

void loop(){
  copier.copy();

}
```
```
``

**Reply by MohammedSamy** [reply]
I had also tried to put the converter between the Url input and the copier as an input conversion, and tried to put it between copier and the decoder as an output converter. still same problem.

**Reply by pschatzmann** [reply]
The FormatConverter works on PCM data, you can't provide it with mp3!
So it only work after the decoding...
